### PR TITLE
feat(eslint-plugin): [sort-type-constituents] support case sensitive sorting

### LINKED ISSUE #4602: [sort-type-constituents] Feature request: support case sensitive sorting
**Repro**

```TS
type T = "DeleteForever" | "DeletedAt";
```

**Expected Result**

With new feature should leave as is.

**Actual Result**

There is currently no way to use case sensitive sorting.

**Additional Info**

1. New option will help make this rule consistent with other sorting-related rules.
2. When you have values consisting of several words (like "DeleteForever", "DeletedAt"), case sensitive sorting is more logical. Compare it to kebabcase: "delete-forever", "deleted-at" is more natural.
3. For case-sensitive language (e.g. typescript) case sensitive sorting is good option.

Also, natural number order may also be turned on/off with an option.

### PR DESCRIPTION
**Pull request by jsfm01**
## PR Checklist

- [x] Addresses an existing open issue: fixes #4602 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

New case sensitive sorting option: 

`{
  "rules": {
    "@typescript-eslint/sort-type-constituents": [
      "error",
      {
        "caseSensitive": true
      }
    ]
  }
}`

New option will help make this rule consistent with other sorting-related rules.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
Note that we're feature-freezing `sort-type-constituents` and other naming & sorting stylistic rules: #8792. Since this PR was sent before #8792 was filed, we'll still accept it. We'd hate to mark work as accepted, then pull it out from under you after you've done the work. 🙂

**Comment by jsfm01** [comment]
Hi @JoshuaKGoldberg 
This PR is ready for review!

File: packages/eslint-plugin/src/rules/sort-type-constituents.ts [hunk 1]
```diff
@@ -132,6 +142,10 @@ export default createRule<Options, MessageIds>({
             description: 'Whether to check union types.',
             type: 'boolean',
           },
+          caseSensitiveSorting: {
```
[the remark this answers is not in the document]
Comment (jsfm01) [reply packages/eslint-plugin/src/rules/sort-type-constituents.ts]: I totally agree with you! Just committed this changes :)

File: packages/eslint-plugin/tests/rules/sort-type-constituents.test.ts
[the remark this answers is not in the document]
Comment (jsfm01) [reply packages/eslint-plugin/tests/rules/sort-type-constituents.test.ts:1]: I just committed two more tests. Please take a look and let me know what you think!

File: packages/eslint-plugin/docs/rules/sort-type-constituents.mdx [hunk 2]
```diff
@@ -97,6 +97,29 @@ type T4 =
 
 ## Options
 
+### `caseSensitive`
+
+Whether to sort using case sensitive.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/sort-type-constituents.mdx:102]:
```suggestion
Whether to sort using case sensitive string comparisons.
```

**Review by JoshuaKGoldberg — APPROVED** [review]
LGTM, thanks! 🙌 

Leaving with `1 approval` in case anyone else wants to take a look.

File: packages/eslint-plugin/src/rules/sort-type-constituents.ts [hunk 3]
```diff
@@ -164,7 +179,10 @@ export default createRule<Options, MessageIds>({
       ],
     },
   ],
-  create(context, [{ checkIntersections, checkUnions, groupOrder }]) {
+  create(
+    context,
+    [{ checkIntersections, checkUnions, caseSensitive, groupOrder }],
+  ) {
     const collator = new Intl.Collator('en', {
       sensitivity: 'base',
```
[the remark this answers is not in the document]
Comment (jsfm01) [reply packages/eslint-plugin/src/rules/sort-type-constituents.ts:187]: When the sensitive case is true, it follows the Unicode code points, no matter the collator sensitivity. If the collator sensitivity is set to 'case', for instance, the sorting would still be case-sensitive, but diacritics and accents would be treated as insignificant. The “accent”, doesn’t distinguish the capital letters, for instance, a = A. I believe we should keep as ‘base’.

**Pull request merged**